File: server/device/device.go
package device

import (
	"net"
	"sync"
	"time"
)

type Status string

const (
	StatusStandby      Status = "standby"
	StatusTransmitting Status = "transmitting"
	StatusReceiving    Status = "receiving"
	StatusDisconnected Status = "disconnected"
)

type Device struct {
	DeviceID    uint32
	SessionID   uint8
	GroupID     uint8
	Status      Status
	Conn        net.Conn
	UDPAddr     *net.UDPAddr
	ConnectedAt time.Time
	LastSeen    time.Time
}

var (
	mu      sync.RWMutex
	devices = make(map[uint8]*Device) // key: session_id
)

// Register は空きセッションID（1〜10）を割り当てて端末を登録する。
// 同一deviceIDが再接続した場合は既存エントリを上書きし、古い net.Conn を返す
// （呼び出し元で Close すること）。満杯の場合は (nil, nil) を返す。
func Register(deviceID uint32, groupID uint8, conn net.Conn) (*Device, net.Conn) {
	mu.Lock()
	defer mu.Unlock()

	// 既登録の同一device_idは上書き（再接続）
	for _, d := range devices {
		if d.DeviceID == deviceID {
			oldConn := d.Conn
			d.GroupID = groupID
			d.Conn = conn
			d.Status = StatusStandby
			d.ConnectedAt = time.Now()
			d.LastSeen = time.Now()
			return d, oldConn
		}
	}

	// 空きセッションID 1〜10 を順番に探す
	for sid := uint8(1); sid <= 10; sid++ {
		if _, occupied := devices[sid]; !occupied {
			d := &Device{
				DeviceID:    deviceID,
				SessionID:   sid,
				GroupID:     groupID,
				Status:      StatusStandby,
				Conn:        conn,
				ConnectedAt: time.Now(),
				LastSeen:    time.Now(),
			}
			devices[sid] = d
			return d, nil
		}
	}
	return nil, nil // 満杯
}

// OwnedBy は指定セッションの現在の Conn が conn と同一かどうかを返す。
// 再接続で置き換えられた古いgoroutineがRemoveを誤って呼ばないためのガード。
func OwnedBy(sessionID uint8, conn net.Conn) bool {
	mu.RLock()
	defer mu.RUnlock()
	d, ok := devices[sessionID]
	return ok && d.Conn == conn
}

// UpdateLastSeen は端末の最終応答時刻を更新する
func UpdateLastSeen(sessionID uint8) {
	mu.Lock()
	defer mu.Unlock()
	if d, ok := devices[sessionID]; ok {
		d.LastSeen = time.Now()
	}
}

// Get は指定セッションIDの端末を返す
func Get(sessionID uint8) *Device {
	mu.RLock()
	defer mu.RUnlock()
	return devices[sessionID]
}

// UpdateStatus は端末のステータスを更新する
func UpdateStatus(sessionID uint8, status Status) {
	mu.Lock()
	defer mu.Unlock()
	if d, ok := devices[sessionID]; ok {
		d.Status = status
	}
}

// UpdateGroup は端末のグループIDを更新する
func UpdateGroup(sessionID uint8, groupID uint8) {
	mu.Lock()
	defer mu.Unlock()
	if d, ok := devices[sessionID]; ok {
		d.GroupID = groupID
	}
}

// AllInGroup は指定グループの接続端末を返す
func AllInGroup(groupID uint8) []*Device {
	mu.RLock()
	defer mu.RUnlock()
	list := make([]*Device, 0)
	for _, d := range devices {
		if d.GroupID == groupID {
			list = append(list, d)
		}
	}
	return list
}

func UpdateUDPAddr(sessionID uint8, addr *net.UDPAddr) {
	mu.Lock()
	defer mu.Unlock()
	if d, ok := devices[sessionID]; ok {
		d.UDPAddr = addr
	}
}

func GetUDPAddr(sessionID uint8) *net.UDPAddr {
	mu.RLock()
	defer mu.RUnlock()
	if d, ok := devices[sessionID]; ok {
		return d.UDPAddr
	}
	return nil
}

func Remove(sessionID uint8) {
	mu.Lock()
	defer mu.Unlock()
	delete(devices, sessionID)
}

func All() []*Device {
	mu.RLock()
	defer mu.RUnlock()
	list := make([]*Device, 0, len(devices))
	for _, d := range devices {
		list = append(list, d)
	}
	return list
}

File: server/device.py
"""
端末セッション管理。
asyncio シングルスレッドで動作するため、ロック不要。
"""
import time
from dataclasses import dataclass, field
from typing import Optional

Addr = tuple[str, int]


@dataclass
class Device:
    device_id:    int
    session_id:   int
    group_id:     int
    status:       str  = "standby"
    ctrl_addr:    Optional[Addr] = None   # 制御UDP送信元
    audio_addr:   Optional[Addr] = None   # 音声UDP送信元
    connected_at: float = field(default_factory=time.time)
    last_seen:    float = field(default_factory=time.time)


_devices:    dict[int,   Device] = {}  # session_id -> Device
_addr_index: dict[Addr,  int]    = {}  # ctrl_addr  -> session_id


# ---------- 登録 / 削除 ----------

def register(device_id: int, group_id: int, ctrl_addr: Addr) -> Optional[Device]:
    """端末を登録する。再接続の場合は既存エントリを更新して返す。満杯なら None。"""
    # 再接続: 同一 device_id を上書き
    for d in _devices.values():
        if d.device_id == device_id:
            if d.ctrl_addr and d.ctrl_addr in _addr_index:
                del _addr_index[d.ctrl_addr]
            d.group_id     = group_id
            d.ctrl_addr    = ctrl_addr
            d.status       = "standby"
            d.connected_at = time.time()
            d.last_seen    = time.time()
            _addr_index[ctrl_addr] = d.session_id
            return d

    # 新規: 空き session_id 1〜10 を探す
    for sid in range(1, 11):
        if sid not in _devices:
            d = Device(device_id=device_id, session_id=sid,
                       group_id=group_id, ctrl_addr=ctrl_addr)
            _devices[sid]          = d
            _addr_index[ctrl_addr] = sid
            return d

    return None  # 満杯


def remove(session_id: int) -> None:
    d = _devices.pop(session_id, None)
    if d and d.ctrl_addr:
        _addr_index.pop(d.ctrl_addr, None)


# ---------- 検索 ----------

def get(session_id: int) -> Optional[Device]:
    return _devices.get(session_id)


def find_by_ctrl_addr(addr: Addr) -> Optional[Device]:
    sid = _addr_index.get(addr)
    return _devices.get(sid) if sid is not None else None


def all_in_group(group_id: int) -> list[Device]:
    return [d for d in _devices.values() if d.group_id == group_id]


def all_devices() -> list[Device]:
    return list(_devices.values())


# ---------- 更新 ----------

def update_last_seen(session_id: int) -> None:
    if d := _devices.get(session_id):
        d.last_seen = time.time()


def update_ctrl_addr(session_id: int, addr: Addr) -> None:
    d = _devices.get(session_id)
    if d is None:
        return
    if d.ctrl_addr and d.ctrl_addr != addr:
        _addr_index.pop(d.ctrl_addr, None)
    d.ctrl_addr        = addr
    _addr_index[addr]  = session_id


def update_audio_addr(session_id: int, addr: Addr) -> None:
    if d := _devices.get(session_id):
        d.audio_addr = addr


def get_audio_addr(session_id: int) -> Optional[Addr]:
    if d := _devices.get(session_id):
        return d.audio_addr
    return None


def update_status(session_id: int, status: str) -> None:
    if d := _devices.get(session_id):
        d.status = status


def update_group(session_id: int, group_id: int) -> None:
    if d := _devices.get(session_id):
        d.group_id = group_id
